src/tmpfiles/tmpfiles.c

```diff
@@ -2003,6 +2005,177 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
+                        (void) log_error_errno(errno, "fstatat(%s, %s): %m", strna(parent_name), name);
+                }
+                return -errno;
+        }
+
+        /* Fail before removing anything if this is an unsafe transition. */
+        /* Ideally this would be fs-util's unsafe_transition, but it is not exported. */
+        if (follow && parent_st->st_uid != 0 && parent_st->st_uid != st.st_uid) {
+                (void) fd_get_path(parent_fd, &parent_name);
+                return log_error_errno(-ENOLINK, "unsafe transition \"%s/%s\": %m", strna(parent_name), name);
+        }
+
+        if ((st.st_mode & S_IFMT) == mode)
+                return 0;
+
+        (void) fd_get_path(parent_fd, &parent_name);
+        log_warning("wrong file type 0x%x; rm -rf \"%s/%s\"", st.st_mode & S_IFMT, strna(parent_name), name);
+
+        /* If the target of the symlink was the wrong type, the link needs to be removed instead of the
+         * target, so make sure it is identified as a link and not a directory. */
+        if (follow) {
+                r = fstatat(parent_fd, name, &st, AT_SYMLINK_NOFOLLOW);
+                if (r < 0) {
+                        if (errno != ENOENT) {
+                                (void) log_error_errno(errno, "fstatat(%s, %s): %m", strna(parent_name), name);
+                        }
+                        return -errno;
+                }
+        }
+
+        /* Do not remove mount points. */
+        if (parent_st->st_dev != st.st_dev) {
+                log_error("mount point at \"%s/%s\"", strna(parent_name), name);
+                return -EBUSY;
+        }
+
+        if ((st.st_mode & S_IFMT) == S_IFDIR) {
+                _cleanup_close_ int child_fd = -1;
+
+                child_fd = openat(parent_fd, name, O_NOCTTY | O_CLOEXEC | O_DIRECTORY);
+                if (child_fd < 0)
+                        return log_error_errno(errno, "failed to open for rm-rf \"%s/%s\": %m", strna(parent_name), name);
+
+                r = rm_rf_children(TAKE_FD(child_fd), REMOVE_ROOT|REMOVE_SUBVOLUME|REMOVE_PHYSICAL, &st);
+                if (r < 0) {
+                        return log_error_errno(r, "failed to remove children \"%s/%s\": %m", strna(parent_name), name);
+                }
+
+                /* Ideally this would be unlinkat_harder but it isn't exported. */
+                r = unlinkat(parent_fd, name, AT_REMOVEDIR);
+        } else {
+                /* Ideally this would be unlinkat_harder but it isn't exported. */
+                r = unlinkat(parent_fd, name, 0);
+        }
+        if (r < 0) {
+                if (errno != ENOENT) {
+                        (void) log_error_errno(errno, "failed to unlink \"%s/%s\": %m", strna(parent_name), name);
+                }
+                return -errno;
+        }
+        return -ENOENT;
+}
+
+static void close_parent(int* parent_fd) {
+        if (*parent_fd != AT_FDCWD) {
+                safe_close(*parent_fd);
+        }
+}
```

Comment:
why? _cleanup_close_ should be enough for this?

Reply:
I switched it over because of an error I had been getting in troubleshooting but the cause of the error was actually the helper function taking ownership of an fd rather than safe_clean not handling AT_FDCWD (which it does).